### PR TITLE
refactor(tls): do not generate tls certificate when server starts up

### PR DESCRIPTION
**Pull request by wujunzhuo**
FYI: @fanweixiao @figroc @venjiang @xiaojian-hong 

---
This change is [Reviewable](https://reviewable.io/reviews/yomorun/yomo/280)

### CODE REVIEW

File: pkg/tls/tls.go [hunk 1]
```diff
@@ -61,31 +54,88 @@ func GenerateCertificate(host ...string) (tls.Certificate, error) {
...
+
+	return &tls.Config{
+		Certificates:       []tls.Certificate{tlsCert},
+		ClientSessionCache: tls.NewLRUClientSessionCache(1),
+		NextProtos:         []string{"yomo"},
+	}, nil
+}
+
+// !!! For test only, do NOT use it in production environment !!!
+func getDevCertAndKey() ([]byte, []byte) {
```
Comment (fanweixiao) [inline pkg/tls/tls.go:124]: share we put this to `.env.local` file?

**Review by fanweixiao — APPROVED** [review]

**Review by figroc — COMMENTED** [review]
建议还是分离更彻底一点，把生成的代码完全去除

File: core/client.go [hunk 2]
```diff
@@ -315,8 +315,9 @@ func (c *Client) initOptions() {
 	}
 	// tls config
 	if c.opts.TLSConfig == nil {
+		env := os.Getenv("YOMO_ENV")
 		c.opts.TLSConfig = &tls.Config{
-			InsecureSkipVerify: true,
+			InsecureSkipVerify: len(env) == 0 || env == "development",
```
[the remark this answers is not in the document]
Comment (fanweixiao) [reply core/client.go:320]: @figroc gotcha 👍🏼

[the remark this answers is not in the document]
Comment (fanweixiao) [reply core/client.go:320]: > 建议还是分离更彻底一点，把生成的代码完全去除

这个咱们是否可以放在 `v2.x` 版本再去改动，因为这个改动会 break 掉已经部署现在这个版本的用户的升级了。

**Comment by fanweixiao** [comment]
👍🏼  
I suggest we merge this to `v1.7`

File: pkg/tls/tls.go [hunk 3]
```diff
@@ -1,91 +1,164 @@
...
-// GenerateTLSConfig Setup a bare-bones TLS config for the server
-func GenerateTLSConfig(host ...string) *tls.Config {
-	tlsCert, _ := GenerateCertificate(host...)
-
-	return &tls.Config{
-		Certificates:       []tls.Certificate{tlsCert},
-		ClientSessionCache: tls.NewLRUClientSessionCache(1),
-		NextProtos:         []string{"yomo"},
+// Create server tls config.
+func CreateServerTLSConfig() (*tls.Config, error) {
```
Comment (fanweixiao) [inline pkg/tls/tls.go:12]: @wujunzhuo should change comment to `CreateServerTLSConfig create server tls config.`

File: pkg/tls/tls.go [hunk 4]
```diff
@@ -1,91 +1,164 @@
...
+		Certificates: []tls.Certificate{*tlsCert},
+		ClientCAs:    pool,
+		ClientAuth:   clientAuth,
+		NextProtos:   []string{"yomo"},
+	}, nil
+}
 
-	serialNumberLimit := new(big.Int).Lsh(big.NewInt(1), 128)
-	serialNumber, err := rand.Int(rand.Reader, serialNumberLimit)
+// Create client tls config.
```
Comment (fanweixiao) [inline pkg/tls/tls.go:38]: @wujunzhuo should change comment to `CreateClientTLSConfig create client tls config.`

**Review by fanweixiao — CHANGES_REQUESTED** [review]

File: test/tls/generate_key.sh [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+  -CA ca.crt -CAkey ca.key -CAcreateserial \
+  -extfile <(printf "subjectAltName=DNS:localhost")
+
+# debug output the certificate
+openssl x509 -noout -text -in cert.crt
+
+# we don't need the CA key, the serial number and the CSR any more
+rm ca.key cert.csr
+
+
```
Comment (fanweixiao) [inline test/tls/generate_key.sh:25]: seems we have to do a lot of documenting job for CA and sort of stuffs to developers. Most of junior developers know little about these.

**Review by fanweixiao — APPROVED** [review]

**Pull request merged**